### PR TITLE
METRON-1333 Ensure that ansible-docker can be used to build metron ( including rpms )

### PR DESCRIPTION
**Pull request by ottobackwards**
The ansible-docker container could be used to build metron ( even the rpms ) and run the ansible scripts. This is no longer true, as the new node ui projects have native compilation that requires c++ 11 support in gcc, and rpm build requires node to be installed ( not local installed as in the mvn packaging ).
Many users are having problems building metron on centos platforms, having this container available and functioning would be a great help.

This PR addresses this by:

- installing devtoolset-4 into the container to get cpp 11
- creating a .bash_profile for root that enables the devtoolset environment
- installing node into the container since building rpms now requires node ( and not just the local node used during building in maven )
- documenting building in container, as well as mapping local .m2 repo into the container

You can now build and test metron, including building the rpms.

## Testing

Follow the steps in the read me and build metron:

`mvn clean package`
`mvn clean package -DskipTests`
`mvn clean install && cd metron-deployment && mvn package -P build-rpms`

Also, after building with the rpms, exit docker
and do full_dev with
`vagrant --ansible-skip-tags="build, quick_dev" up`  and verify that you can use vagrant with the doc built packaging and rpms.

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x ] Have you run a build, test, build rpm set of builds in the container?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book?

### CODE REVIEW

File: metron-deployment/packaging/docker/ansible-docker/Dockerfile [hunk 1]
```diff
@@ -14,13 +14,18 @@
...
-FROM centos:centos6
+FROM centos:centos6.9
 MAINTAINER Apache Metron
 
 RUN yum install -y tar
 RUN yum install -y wget
+# base development tools required
 RUN yum groupinstall -y "Development tools"
+# newer cpp 11 support required for building node modules
+RUN wget http://people.centos.org/tru/devtools-2/devtools-2.repo -O /etc/yum.repos.d/devtools-2.repo
```
Comment (JonZeolla) [inline metron-deployment/packaging/docker/ansible-docker/Dockerfile]: Please upgrade to devtoolset-4.  
```
yum -y install centos-release-scl
yum -y install devtoolset-4-gcc devtoolset-4-gcc-c++ devtoolset-4-binutils
```
or similar.

File: metron-deployment/packaging/docker/ansible-docker/Dockerfile [hunk 2]
```diff
@@ -33,18 +38,28 @@ RUN tar xvf setuptools-11.3.tar.gz
...
 RUN wget http://apache.cs.utah.edu/maven/maven-3/3.3.9/binaries/apache-maven-3.3.9-bin.tar.gz
 RUN tar xzvf apache-maven-3.3.9-bin.tar.gz
 RUN mv apache-maven-3.3.9 /opt/maven
 RUN ln -s /opt/maven/bin/mvn /usr/bin/mvn
-RUN yum -y install asciidoc rpm-build rpm2cpio tar unzip xmlto zip rpmlint && yum clean all
+# install rpm tools required to build rpms
+RUN yum -y install asciidoc rpm-build rpm2cpio tar unzip xmlto zip rpmlint make && yum clean all
+# create a .bashrc for root, enabling the cpp 11 toolset
+RUN touch /root/.bashrc \
+ && cat '/opt/rh/devtoolset-2/enable' >> /root/.bashrc
```
Comment (JonZeolla) [inline metron-deployment/packaging/docker/ansible-docker/Dockerfile]: Same as above

File: metron-deployment/packaging/docker/ansible-docker/README.md [hunk 3]
```diff
@@ -1,17 +1,34 @@
 # Overview
-The Metron ansible-docker container is provided in an effort reduce the installation burden of deploying Metron in a live envirionment.
-It is provisioned with software required to sucessfully run the deployment scripts.
+The Metron ansible-docker container is provided in an effort reduce the installation burden of building Metron.
+It may also be used to deploy Metron in a live environment.
+It is provisioned with software required to sucessfully build metron run the deployment scripts.
```
Comment (JonZeolla) [inline metron-deployment/packaging/docker/ansible-docker/README.md]: `s/metron run/metron and run/`

File: metron-deployment/packaging/docker/ansible-docker/Dockerfile (hunk 1 above)
Comment (ottobackwards) [reply metron-deployment/packaging/docker/ansible-docker/Dockerfile]: yeah, thanks.  Nick also pointed this out in a jira ( referring to your steps ), i'll have this in in a moment

File: metron-deployment/packaging/docker/ansible-docker/README.md (hunk 3 above)
Comment (ottobackwards) [reply metron-deployment/packaging/docker/ansible-docker/README.md]: thanks!

**Comment by ottobackwards** [comment]
Added above, adding as a comment to make sure it is seen

---------

Also, after building with the rpms, exit docker
and do full_dev with
`vagrant --ansible-skip-tags="build" up`  and verify that you can use vagrant with the doc built packaging and rpms.

**Comment by JonZeolla** [comment]
Tested via
```
cd /Users/jzeolla/metron-pr848
docker run -it -v /Users/jzeolla/metron-pr848:/root/metron ansible-docker:2.0.0.2 bash
cd /root/metron
mvn clean package -DskipTests # Failure
```

Ran into
```
[ERROR] npm ERR! Linux 4.9.49-moby
[ERROR] npm ERR! argv "/root/metron/metron-interface/metron-config/node/node" "/root/metron/metron-interface/metron-config/node/node_modules/npm/bin/npm-cli.js" "run" "build"
[ERROR] npm ERR! node v6.2.0
[ERROR] npm ERR! npm  v3.8.9
[ERROR] npm ERR! code ELIFECYCLE
[ERROR] npm ERR! metron-management-ui@0.4.2 build: `./node_modules/angular-cli/bin/ng build -prod`
[ERROR] npm ERR! Exit status 1
[ERROR] npm ERR!
[ERROR] npm ERR! Failed at the metron-management-ui@0.4.2 build script './node_modules/angular-cli/bin/ng build -prod'.
[ERROR] npm ERR! Make sure you have the latest version of node.js and npm installed.
[ERROR] npm ERR! If you do, this is most likely a problem with the metron-management-ui package,
[ERROR] npm ERR! not with npm itself.
[ERROR] npm ERR! Tell the author that this fails on your system:
[ERROR] npm ERR!     ./node_modules/angular-cli/bin/ng build -prod
[ERROR] npm ERR! You can get information on how to open an issue for this project with:
[ERROR] npm ERR!     npm bugs metron-management-ui
[ERROR] npm ERR! Or if that isn't available, you can get their info via:
[ERROR] npm ERR!     npm owner ls metron-management-ui
[ERROR] npm ERR! There is likely additional logging output above.
[ERROR]
[ERROR] npm ERR! Please include the following file with any support request:
[ERROR] npm ERR!     /root/metron/metron-interface/metron-config/npm-debug.log
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Metron ............................................. SUCCESS [ 18.059 s]
[INFO] metron-stellar ..................................... SUCCESS [  7.965 s]
[INFO] stellar-common ..................................... SUCCESS [01:27 min]
[INFO] metron-analytics ................................... SUCCESS [  0.012 s]
[INFO] metron-maas-common ................................. SUCCESS [ 10.648 s]
[INFO] metron-platform .................................... SUCCESS [  0.031 s]
[INFO] metron-zookeeper ................................... SUCCESS [  1.542 s]
[INFO] metron-test-utilities .............................. SUCCESS [ 42.995 s]
[INFO] metron-integration-test ............................ SUCCESS [ 40.759 s]
[INFO] metron-maas-service ................................ SUCCESS [ 10.431 s]
[INFO] metron-common ...................................... SUCCESS [ 46.535 s]
[INFO] metron-statistics .................................. SUCCESS [ 22.078 s]
[INFO] metron-writer ...................................... SUCCESS [ 33.132 s]
[INFO] metron-storm-kafka-override ........................ SUCCESS [  4.151 s]
[INFO] metron-storm-kafka ................................. SUCCESS [  2.083 s]
[INFO] metron-hbase ....................................... SUCCESS [ 16.480 s]
[INFO] metron-profiler-common ............................. SUCCESS [  5.086 s]
[INFO] metron-profiler-client ............................. SUCCESS [ 26.173 s]
[INFO] metron-profiler .................................... SUCCESS [ 56.438 s]
[INFO] metron-hbase-client ................................ SUCCESS [ 16.283 s]
[INFO] metron-enrichment .................................. SUCCESS [01:01 min]
[INFO] metron-indexing .................................... SUCCESS [ 22.402 s]
[INFO] metron-solr ........................................ SUCCESS [ 56.301 s]
[INFO] metron-pcap ........................................ SUCCESS [  5.602 s]
[INFO] metron-parsers ..................................... SUCCESS [01:09 min]
[INFO] metron-pcap-backend ................................ SUCCESS [ 35.876 s]
[INFO] metron-data-management ............................. SUCCESS [01:46 min]
[INFO] metron-api ......................................... SUCCESS [ 57.116 s]
[INFO] metron-management .................................. SUCCESS [ 14.077 s]
[INFO] elasticsearch-shaded ............................... SUCCESS [ 11.380 s]
[INFO] metron-elasticsearch ............................... SUCCESS [01:02 min]
[INFO] metron-deployment .................................. SUCCESS [  0.006 s]
[INFO] Metron Ambari Management Pack ...................... SUCCESS [  4.145 s]
[INFO] metron-contrib ..................................... SUCCESS [  0.015 s]
[INFO] metron-docker ...................................... SUCCESS [  5.673 s]
[INFO] metron-interface ................................... SUCCESS [  0.005 s]
[INFO] metron-config ...................................... FAILURE [04:55 min]
[INFO] metron-alerts ...................................... SKIPPED
[INFO] metron-rest-client ................................. SKIPPED
[INFO] metron-rest ........................................ SKIPPED
[INFO] site-book .......................................... SKIPPED
[INFO] 3rd party Functions (just for tests) ............... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 21:01 min
[INFO] Finished at: 2017-11-28T01:34:36+00:00
[INFO] Final Memory: 154M/394M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.3:npm (ng build) on project metron-config: Failed to run task: 'npm run build' failed. (error code 1) -> [Help 1]
```

More detail

```
0 info it worked if it ends with ok
1 verbose cli [ '/root/metron/metron-interface/metron-config/node/node',
1 verbose cli   '/root/metron/metron-interface/metron-config/node/node_modules/npm/bin/npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'build' ]
2 info using npm@3.8.9
3 info using node@v6.2.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle metron-management-ui@0.4.2~prebuild: metron-management-ui@0.4.2
6 silly lifecycle metron-management-ui@0.4.2~prebuild: no script for prebuild, continuing
7 info lifecycle metron-management-ui@0.4.2~build: metron-management-ui@0.4.2
8 verbose lifecycle metron-management-ui@0.4.2~build: unsafe-perm in lifecycle true
9 verbose lifecycle metron-management-ui@0.4.2~build: PATH: /root/metron/metron-interface/metron-config/node/node_modules/npm/bin/node-gyp-bin:/root/metron/metron-interface/metron-config/node_modules/.bin:/root/metron/metron-interface/metron-config/node:/root/metron/metron-interface/metron-config/node:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:
10 verbose lifecycle metron-management-ui@0.4.2~build: CWD: /root/metron/metron-interface/metron-config
11 silly lifecycle metron-management-ui@0.4.2~build: Args: [ '-c', './node_modules/angular-cli/bin/ng build -prod' ]
12 silly lifecycle metron-management-ui@0.4.2~build: Returned: code: 1  signal: null
13 info lifecycle metron-management-ui@0.4.2~build: Failed to exec build script
14 verbose stack Error: metron-management-ui@0.4.2 build: `./node_modules/angular-cli/bin/ng build -prod`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/root/metron/metron-interface/metron-config/node/node_modules/npm/lib/utils/lifecycle.js:245:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/root/metron/metron-interface/metron-config/node/node_modules/npm/lib/utils/spawn.js:24:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:850:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:5)
15 verbose pkgid metron-management-ui@0.4.2
16 verbose cwd /root/metron/metron-interface/metron-config
17 error Linux 4.9.49-moby
18 error argv "/root/metron/metron-interface/metron-config/node/node" "/root/metron/metron-interface/metron-config/node/node_modules/npm/bin/npm-cli.js" "run" "build"
19 error node v6.2.0
20 error npm  v3.8.9
21 error code ELIFECYCLE
22 error metron-management-ui@0.4.2 build: `./node_modules/angular-cli/bin/ng build -prod`
22 error Exit status 1
23 error Failed at the metron-management-ui@0.4.2 build script './node_modules/angular-cli/bin/ng build -prod'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the metron-management-ui package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     ./node_modules/angular-cli/bin/ng build -prod
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs metron-management-ui
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls metron-management-ui
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

**Comment by JonZeolla** [comment]
+1, this is awesome work thanks Otto.  Tested via
```
cd metron-deployment/packaging/docker/ansible-docker/
docker build -t ansible-docker:2.0.0.2 .
docker run -it -v /Users/jzeolla/metron-pr848:/root/metron ansible-docker:2.0.0.2 bash
cd /root/metron
mvn clean package -DskipTests # Success
mvn clean install && cd metron-deployment && mvn package -P build-rpms # Success
exit
cd ../../../vagrant/full-dev-platform/
vagrant --ansible-skip-tags="build,quick_dev" up # Success
```

**Comment by ottobackwards** [comment]
@JonZeolla I would like to change the readme example to include the quick_dev skip.

**Comment by JonZeolla** [comment]
Ahh sorry, missed that was in the docs as well.  Ok, looks good

**Pull request closed without merging**